src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.AddDebuggerDisplay
+{
+    internal abstract class AbstractAddDebuggerDisplayCodeRefactoringProvider<TTypeDeclarationSyntax, TMethodDeclarationSyntax> : CodeRefactoringProvider
+        where TTypeDeclarationSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+    {
+        private const string DebuggerDisplayMethodName = "GetDebuggerDisplay";
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var typeAndPriority =
+                await GetRelevantTypeFromHeaderAsync(context).ConfigureAwait(false)
+                ?? await GetRelevantTypeFromMethodAsync(context).ConfigureAwait(false);
+
+            if (!(typeAndPriority is var (type, priority))) return;
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var typeSymbol = (ITypeSymbol)semanticModel.GetDeclaredSymbol(type, context.CancellationToken);
+
+            if (IsClassOrStruct(typeSymbol) && !HasDebuggerDisplayAttribute(typeSymbol, semanticModel.Compilation))
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    priority,
+                    FeaturesResources.Add_DebuggerDisplay,
+                    cancellationToken => ApplyAsync(context.Document, type, cancellationToken)));
+            }
+        }
+
+        private static async Task<(TTypeDeclarationSyntax Type, CodeActionPriority Priority)?> GetRelevantTypeFromHeaderAsync(CodeRefactoringContext context)
```

Comment:
📝 We use camelCase for tuple elements in the IDE layers (workspaces, features, editor features).